### PR TITLE
graduate CSIServiceAccountToken to beta

### PR DESCRIPTION
**Pull request by zshihang**
#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:
graduate CSIServiceAccountToken to beta

#### Does this PR introduce a user-facing change?
```release-note
CSIServiceAccountToken is Beta now
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1855-csi-driver-service-account-token
```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -639,6 +639,7 @@ const (
 
 	// owner: @zshihang
 	// alpha: v1.20
+	// beta: v1.21
```
Comment (msau42) [inline pkg/features/kube_features.go:642]: A couple missing things:
- Update the comments in types.go indicating the API fields are beta.
- Update the e2e test to remove the "Feature" tag.

Comment (zshihang) [reply pkg/features/kube_features.go:643]: done. thanks.

File: pkg/apis/storage/types.go [hunk 2]
```diff
@@ -363,11 +363,11 @@ type CSIDriverSpec struct {
 	// most one token is empty string. To receive a new token after expiry,
 	// RequiresRepublish can be used to trigger NodePublishVolume periodically.
 	//
-	// This is an alpha feature and only available when the
```
Comment (msau42) [inline pkg/apis/storage/types.go:366]: Can you keep the comment but change it to beta?

Comment (zshihang) [reply pkg/apis/storage/types.go:366]: beta is enabled by default right?

Comment (msau42) [reply pkg/apis/storage/types.go:366]: It is enabled by default, but it could be disabled by a cluster operator. I think we mainly want to inform users that this is still a beta feature so they shouldn't depend on the API/feature being stable yet.

Comment (zshihang) [reply pkg/apis/storage/types.go:366]: makes sense, done.

**Comment by zshihang** [comment]
/assign @liggitt
Jordan, could you do the API review?

File: pkg/apis/storage/types.go [hunk 3]
```diff
@@ -363,11 +363,14 @@ type CSIDriverSpec struct {
...
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
+	// +patchMergeKey=audience
+	// +patchStrategy=merge
+	// +listMapKey=audience
+	// +listType=map
 	TokenRequests []TokenRequest
```
Comment (liggitt) [inline pkg/apis/storage/types.go:374]: do you have a reference for why this change is being made? do we expect multiple actors to coordinate on this field or to specify the entire list at once?

File: staging/src/k8s.io/api/storage/v1/generated.proto [hunk 4]
```diff
@@ -163,11 +163,14 @@ message CSIDriverSpec {
...
-  // This is an alpha feature and only available when the
+  // This is a beta feature and only available when the
   // CSIServiceAccountToken feature is enabled.
   //
   // +optional
-  // +listType=atomic
+  // +patchMergeKey=audience
+  // +patchStrategy=merge
+  // +listMapKey=audience
+  // +listType=map
```
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/generated.proto:173]: same question about why the listType was changed

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 5]
```diff
@@ -361,12 +361,15 @@ type CSIDriverSpec struct {
...
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
-	TokenRequests []TokenRequest `json:"tokenRequests,omitempty" protobuf:"bytes,6,opt,name=tokenRequests"`
+	// +patchMergeKey=audience
+	// +patchStrategy=merge
+	// +listMapKey=audience
+	// +listType=map
```
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/types.go:371]: same question about why the listType was changed

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 6]
```diff
@@ -361,12 +361,15 @@ type CSIDriverSpec struct {
...
 	//
-	// This is an alpha feature and only available when the
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
-	TokenRequests []TokenRequest `json:"tokenRequests,omitempty" protobuf:"bytes,6,opt,name=tokenRequests"`
+	// +patchMergeKey=audience
+	// +patchStrategy=merge
```
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/types.go:369]: @apelisse can you confirm that changing patchMergeKey / patchStrategy is not compatible with client-side apply?

File: pkg/apis/storage/types.go (hunk 3 above)
Comment (zshihang) [reply pkg/apis/storage/types.go:374]: in alpha, we didn't use `patchStrategy` because `patchMergeKey` doesn't support empty string in 1.20. talked to @apelisse, that is supported in 1.21 now.

yea, there could be multiple actors on this field.

Comment (liggitt) [reply pkg/apis/storage/types.go:374]: The issue is that patchStrategy/patchMergeKey gets baked into clients and affects the patch kubectl sends when invoking `kubectl apply`. Changing that here means a 1.20 kubectl will send the wrong content when updating this field via apply.

Comment (apelisse) [reply pkg/apis/storage/types.go:374]: Doesn't kubectl uses the openapi at that point to make the merge decision?

File: staging/src/k8s.io/api/storage/v1/types.go (hunk 5 above)
Comment (apelisse) [reply staging/src/k8s.io/api/storage/v1/types.go:371]: @jpbetz has worked on making it possible to change the listType.

Comment (jpbetz) [reply staging/src/k8s.io/api/storage/v1/types.go:371]: Specifically: https://github.com/kubernetes-sigs/structured-merge-diff/pull/170

File: pkg/apis/storage/types.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/storage/types.go:374]: can you verify that is true (e.g. use a 1.20 kubectl to client-side apply a change to a server running with this PR and see if it honors the new patch strategy)

Comment (zshihang) [reply pkg/apis/storage/types.go:374]:
```
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:12:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.908+d2922316ed8131", GitCommit:"d2922316ed81319336024e9368f3ffb6ce6d981c", GitTreeState:"clean", BuildDate:"2021-03-04T18:07:21Z", GoVersion:"go1.16", Compiler:"gc", Platform:"linux/amd64"}
```

the following is the log when i am trying to apply a CSIDriver object(replaced audience "gcp" with "newgcp"). the new patch strategy is used and the calculation is correct. however, it failed to do so due to "immutable field" while the `TokenRequests` field is mutable.
```
I0304 10:33:17.444489 3247191 round_trippers.go:425] curl -k -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Authorization: Bearer <masked>" 'https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com'
I0304 10:33:17.455476 3247191 round_trippers.go:445] GET https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com 200 OK in 10 milliseconds
I0304 10:33:17.455516 3247191 round_trippers.go:451] Response Headers:
I0304 10:33:17.455531 3247191 round_trippers.go:454]     Date: Thu, 04 Mar 2021 18:33:17 GMT
I0304 10:33:17.455544 3247191 round_trippers.go:454]     Audit-Id: d88cdc8b-5985-450e-997c-e01a5cf0afb0
I0304 10:33:17.455554 3247191 round_trippers.go:454]     Cache-Control: no-cache, private
I0304 10:33:17.455565 3247191 round_trippers.go:454]     Content-Type: application/json
I0304 10:33:17.455578 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Flowschema-Uid: 66e63b9b-77dd-49b9-94f4-660891cf396c
I0304 10:33:17.455589 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Prioritylevel-Uid: 90666279-523f-47fc-9a23-dc12a60a3632
I0304 10:33:17.455617 3247191 round_trippers.go:454]     Content-Length: 1463
I0304 10:33:17.455695 3247191 request.go:1107] Response Body: {"kind":"CSIDriver","apiVersion":"storage.k8s.io/v1","metadata":{"name":"test.example.com","uid":"1c69bda6-18ea-4ed9-b552-acb56dc57d28","resourceVersion":"1703","creationTimestamp":"2021-03-04T18:31:08Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"gcp\"},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"managedFields":[{"manager":"kubectl-client-side-apply","operation":"Update","apiVersion":"storage.k8s.io/v1","time":"2021-03-04T18:31:08Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{}}},"f:spec":{"f:attachRequired":{},"f:fsGroupPolicy":{},"f:podInfoOnMount":{},"f:requiresRepublish":{},"f:tokenRequests":{".":{},"k:{\"audience\":\"\"}":{".":{},"f:audience":{},"f:expirationSeconds":{}},"k:{\"audience\":\"gcp\"}":{".":{},"f:audience":{}}},"f:volumeLifecycleModes":{".":{},"v:\"Ephemeral\"":{},"v:\"Persistent\"":{}}}}}]},"spec":{"attachRequired":true,"podInfoOnMount":true,"volumeLifecycleModes":["Persistent","Ephemeral"],"fsGroupPolicy":"ReadWriteOnceWithFSType","tokenRequests":[{"audience":"gcp"},{"audience":"","expirationSeconds":3600}],"requiresRepublish":false}}
I0304 10:33:17.457010 3247191 request.go:1107] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"newgcp\"},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"}},"spec":{"$setElementOrder/tokenRequests":[{"audience":"newgcp"},{"audience":""}],"tokenRequests":[{"audience":"newgcp"},{"$patch":"delete","audience":"gcp"}]}}
I0304 10:33:17.457147 3247191 round_trippers.go:425] curl -k -v -XPATCH  -H "Accept: application/json" -H "Content-Type: application/strategic-merge-patch+json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Authorization: Bearer <masked>" 'https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply'
I0304 10:33:17.472575 3247191 round_trippers.go:445] PATCH https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply 422 Unprocessable Entity in 15 milliseconds
I0304 10:33:17.472606 3247191 round_trippers.go:451] Response Headers:
I0304 10:33:17.472622 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Flowschema-Uid: 66e63b9b-77dd-49b9-94f4-660891cf396c
I0304 10:33:17.472639 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Prioritylevel-Uid: 90666279-523f-47fc-9a23-dc12a60a3632
I0304 10:33:17.472655 3247191 round_trippers.go:454]     Content-Length: 1394
I0304 10:33:17.472670 3247191 round_trippers.go:454]     Date: Thu, 04 Mar 2021 18:33:17 GMT
I0304 10:33:17.472686 3247191 round_trippers.go:454]     Audit-Id: bed10d8a-fb41-446e-ad32-33edcd760dce
I0304 10:33:17.472701 3247191 round_trippers.go:454]     Cache-Control: no-cache, private
I0304 10:33:17.472717 3247191 round_trippers.go:454]     Content-Type: application/json
I0304 10:33:17.473462 3247191 request.go:1107] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"CSIDriver.storage.k8s.io \"test.example.com\" is invalid: spec: Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{\"Persistent\", \"Ephemeral\"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:\"newgcp\", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:\"\", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable","reason":"Invalid","details":{"name":"test.example.com","group":"storage.k8s.io","kind":"CSIDriver","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{\"Persistent\", \"Ephemeral\"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:\"newgcp\", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:\"\", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable","field":"spec"}]},"code":422}
F0304 10:33:17.474004 3247191 helpers.go:115] The CSIDriver "test.example.com" is invalid: spec: Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{"Persistent", "Ephemeral"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:"newgcp", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:"", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable
```

Comment (zshihang) [reply pkg/apis/storage/types.go:388]: `CSIDriverSpec` was immutable. changing two fields added by this feature gate to mutable. it is safe to make these two mutable because every `NodePublishVolume` call would read the value. cc @msau42 

the patch strategy is working according to the above log. would test it again with the new changes

Comment (zshihang) [reply pkg/apis/storage/types.go:388]:
```
I0304 16:53:27.994030 3587476 request.go:1107] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"gcp\",\"expirationSeconds\":7200},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"}},"spec":{"$setElementOrder/tokenRequests":[{"audience":"gcp"},{"audience":""}],"tokenRequests":[{"audience":"gcp","expirationSeconds":7200}]}}
I0304 16:53:27.994123 3587476 round_trippers.go:425] curl -k -v -XPATCH  -H "Accept: application/json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Content-Type: application/strategic-merge-patch+json" -H "Authorization: Bearer <masked>" 'https://34.82.67.178/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply'
I0304 16:53:28.010485 3587476 round_trippers.go:445] PATCH https://34.82.67.178/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply 200 OK in 16 milliseconds
```

looks working :)

File: pkg/apis/storage/validation/validation.go [hunk 7]
```diff
@@ -409,10 +409,21 @@ func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
 func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&new.ObjectMeta, &old.ObjectMeta, field.NewPath("metadata"))
 
-	// Spec is read-only
-	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
-	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	// immutable fields should not be mutated.
```
Comment (msau42) [inline pkg/apis/storage/validation/validation.go:412]: should we also call validateCSIDriverSpec again so that we check the updated fields are valid?

File: pkg/registry/storage/csidriver/strategy.go [hunk 8]
```diff
@@ -82,31 +83,33 @@ func (csiDriverStrategy) AllowCreateOnUpdate() bool {
...
+	if oldCSIDriver.Spec.RequiresRepublish == nil &&
 		!utilfeature.DefaultFeatureGate.Enabled(features.CSIServiceAccountToken) {
-		csiDriver := obj.(*storage.CSIDriver)
-		csiDriver.Spec.RequiresRepublish = nil
+		newCSIDriver.Spec.RequiresRepublish = nil
+	}
+
+	// Any changes to the mutable fields increment the generation number.
+	if !reflect.DeepEqual(oldCSIDriver.Spec.TokenRequests, newCSIDriver.Spec.TokenRequests) || !reflect.DeepEqual(oldCSIDriver.Spec.RequiresRepublish, newCSIDriver.Spec.RequiresRepublish) {
+		newCSIDriver.Generation = oldCSIDriver.Generation + 1
```
Comment (msau42) [inline pkg/registry/storage/csidriver/strategy.go:112]: I don't see this being done for any other mutable fields in this group and for other core types like Pod. Is this necessary?

Comment (zshihang) [reply pkg/registry/storage/csidriver/strategy.go:112]: this is mentioned by the removed comment above. honestly i am not sure if this is necessary. there are kinds using `Generation` e.g. `networking/ingressclass/strategy.go`

File: pkg/apis/storage/validation/validation.go (hunk 7 above)
Comment (zshihang) [reply pkg/apis/storage/validation/validation.go:412]: done.

**Comment by zshihang** [comment]
/retest

not sure why pull-kubernetes-verify is falling due to openspec but ` ./hack/verify-openapi-spec.sh` passe locally

File: pkg/registry/storage/csidriver/strategy.go (hunk 8 above)
Comment (zshihang) [reply pkg/registry/storage/csidriver/strategy.go:109]: it seems an informational field and as long as it is not decrementing, then it is fine. i.e. unset or increment. 

@liggitt do you have more to add about how this field is used?

**Comment by msau42** [comment]
This lgtm, there is one open question on whether we need to increment the generation number in the strategy. @liggitt can you do another pass?

File: pkg/apis/storage/types.go [hunk 9]
```diff
@@ -363,11 +377,14 @@ type CSIDriverSpec struct {
 	// most one token is empty string. To receive a new token after expiry,
 	// RequiresRepublish can be used to trigger NodePublishVolume periodically.
 	//
-	// This is an alpha feature and only available when the
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
+	// +patchMergeKey=audience
```
Comment (liggitt) [inline pkg/apis/storage/types.go:384]: I don't think a mergeable list is what we want here (and it actually doesn't work consistently in my experiments).

The audience field is not actually required by the API server (you can `kubectl apply -f ... --validate=false` a manifest that is missing it, or create one via a controller without explicitly sending an audience field in a tokenrequest item)

If you do that, you end up with an existing persisted object `kubectl apply` and `kubectl apply --server-side` don't know how to merge updates to properly.

They either complain and fail:
```
kubectl apply -f ~/snippets/csidriver/driver2.yaml
warning: error calculating patch from openapi spec: map: map[] does not contain declared merge key: audience
error: error when applying patch:

to:
Resource: "storage.k8s.io/v1, Resource=csidrivers", GroupVersionKind: "storage.k8s.io/v1, Kind=CSIDriver"
Name: "tokens", Namespace: ""
for: "/Users/liggitt/snippets/csidriver/driver2.yaml": error when creating patch with:
original:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{},"name":"tokens"},"spec":{"tokenRequests":[{}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

modified:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"tokens\"},\"spec\":{\"tokenRequests\":[{\"audience\":\"another\"}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"name":"tokens"},"spec":{"tokenRequests":[{"audience":"another"}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

current:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"tokens\"},\"spec\":{\"tokenRequests\":[{}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"creationTimestamp":"2021-03-09T20:57:24Z","managedFields":[{"apiVersion":"storage.k8s.io/v1","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{}}},"f:spec":{"f:attachRequired":{},"f:fsGroupPolicy":{},"f:podInfoOnMount":{},"f:requiresRepublish":{},"f:storageCapacity":{},"f:tokenRequests":{".":{},"k:{\"audience\":\"\"}":{".":{},"f:audience":{}}},"f:volumeLifecycleModes":{".":{},"v:\"Ephemeral\"":{},"v:\"Persistent\"":{}}}},"manager":"kubectl-client-side-apply","operation":"Update","time":"2021-03-09T20:57:24Z"}],"name":"tokens","resourceVersion":"330","uid":"98c2d9e1-817f-40a0-b2ba-fae511f4adb1"},"spec":{"attachRequired":true,"fsGroupPolicy":"ReadWriteOnceWithFSType","podInfoOnMount":false,"requiresRepublish":false,"storageCapacity":false,"tokenRequests":[{"audience":""}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

for: "/Users/liggitt/snippets/csidriver/driver2.yaml": map: map[] does not contain declared merge key: audience
```

or silently keep the existing `audience:""` entry and add another item to the list.

I would expect a single actor to be defining the CSIDriver spec and dictating the entire content of this list, which would mean it should stay atomic.

Comment (liggitt) [reply pkg/apis/storage/types.go:384]: `kubectl apply --validate=false` with this:
```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: tokens
spec:
  volumeLifecycleModes:
  - Persistent
  - Ephemeral
  tokenRequests: [{}]
```

then try to `kubectl apply` with this:
```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: tokens
spec:
  volumeLifecycleModes:
  - Persistent
  - Ephemeral
  tokenRequests: [{"audience":"another"}]
```

`kubectl apply --server-side` submits the request, but produces a CSIDriver containing the old audience as well, rather than dropping it properly:
```yaml
  tokenRequests:
  - audience: another
  - audience: ""
```

File: pkg/registry/storage/csidriver/strategy.go [hunk 10]
```diff
@@ -82,31 +80,33 @@ func (csiDriverStrategy) AllowCreateOnUpdate() bool {
...
-	if old.(*storage.CSIDriver).Spec.RequiresRepublish == nil &&
+	if oldCSIDriver.Spec.RequiresRepublish == nil &&
 		!utilfeature.DefaultFeatureGate.Enabled(features.CSIServiceAccountToken) {
-		csiDriver := obj.(*storage.CSIDriver)
-		csiDriver.Spec.RequiresRepublish = nil
+		newCSIDriver.Spec.RequiresRepublish = nil
+	}
+
+	// Any changes to the mutable fields increment the generation number.
+	if !reflect.DeepEqual(oldCSIDriver.Spec.TokenRequests, newCSIDriver.Spec.TokenRequests) || !reflect.DeepEqual(oldCSIDriver.Spec.RequiresRepublish, newCSIDriver.Spec.RequiresRepublish) {
```
Comment (liggitt) [inline pkg/registry/storage/csidriver/strategy.go:108]: use `apiequality.Semantic.DeepEqual`, tokenrequests `null` and `[]` are equivalent, but reflect.DeepEqual considers them different

File: pkg/registry/storage/csidriver/strategy.go (hunk 8 above)
Comment (liggitt) [reply pkg/registry/storage/csidriver/strategy.go:109]: Incrementing generation is typically done for spec changes on objects which are reconciled by controllers and have a `status.observedGeneration` field, but it's fine to do this here as well.

File: pkg/registry/storage/csidriver/strategy.go [hunk 11]
```diff
@@ -64,10 +65,7 @@ func (csiDriverStrategy) PrepareForCreate(ctx context.Context, obj runtime.Objec
 func (csiDriverStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
 	csiDriver := obj.(*storage.CSIDriver)
 
-	errs := validation.ValidateCSIDriver(csiDriver)
-	errs = append(errs, validation.ValidateCSIDriver(csiDriver)...)
-
-	return errs
+	return validation.ValidateCSIDriver(csiDriver)
```
Comment (liggitt) [inline pkg/registry/storage/csidriver/strategy.go:68]: good catch

File: pkg/apis/storage/validation/validation.go [hunk 12]
```diff
@@ -409,11 +409,24 @@ func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
...
+	if !apiequality.Semantic.DeepEqual(old.Spec.AttachRequired, new.Spec.AttachRequired) {
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go:413]: can do all of these as a one-liner like this:
```go
allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(new.Spec.AttachRequired, old.Spec.AttachRequired, field.NewPath("spec", "attachedRequired"))...)
```

File: pkg/apis/storage/validation/validation.go (hunk 7 above)
Comment (liggitt) [reply pkg/apis/storage/validation/validation.go:412]: typically, we try to only validate fields we allow you to change (TokenRequests, I think). if there was an invalid field we also required to be immutable, they'd be stuck without a way to update (updating triggers immutable error, not updating triggers invalid error)

File: pkg/apis/storage/validation/validation.go (hunk 12 above)
Comment (zshihang) [reply pkg/apis/storage/validation/validation.go:413]: good point

File: pkg/registry/storage/csidriver/strategy.go (hunk 10 above)
Comment (zshihang) [reply pkg/registry/storage/csidriver/strategy.go:108]: done

File: pkg/apis/storage/types.go (hunk 9 above)
Comment (zshihang) [reply pkg/apis/storage/types.go]: didn't know there is a `--validate=false` to ignore the validation. how to mark a field as required if`--validate=false`? then i think above case would be prevented.

i was thinking there might be a case where a controller is reconciling the CSIDriver object but yea, single actor is most common. 

changed back to atomic.

File: pkg/apis/storage/validation/validation.go [hunk 13]
```diff
@@ -425,11 +425,14 @@ func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
 func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&new.ObjectMeta, &old.ObjectMeta, field.NewPath("metadata"))
 
-	// Spec is read-only
-	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
-	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
-	}
+	allErrs = append(allErrs, validateCSIDriverSpec(&new.Spec, field.NewPath("spec"))...)
```
Comment (msau42) [inline pkg/apis/storage/validation/validation.go:428]: From https://github.com/kubernetes/kubernetes/pull/99298#discussion_r590742269, the recommendation is that we only validate the mutable fields instead of the whole spec.

File: pkg/apis/storage/validation/validation.go (hunk 7 above)
Comment (zshihang) [reply pkg/apis/storage/validation/validation.go:428]: done

**Comment by zshihang** [comment]
@annajung exception is filed.

**Comment by liggitt** [comment]
/approve

API changes lgtm

**Comment by palnabarun** [comment]
/milestone v1.21

since the exception request was APPROVED. ref: https://groups.google.com/g/kubernetes-sig-release/c/xnR5k3nI6Dk/m/VXbsxE7nAgAJ
